Source/WebGPU/WGSL/TypeCheck.cpp

```diff
@@ -0,0 +1,321 @@
+struct Type {
+    enum Kind {
+        Variable,
+        Primitive,
+        Constructor,
+        Application,
+        Function,
+    };
+};
+
+using SubstituionCallback = std::function<void(Type*)>;
+
+class TypeChecker : public AST::Visitor, public ContextProvider<Type*> {
+    friend class UnificationContext;
+
+public:
+    TypeChecker(AST::ShaderModule& shaderModule)
+        : m_shaderModule(shaderModule)
+    {
+    }
+
+    void check();
+
+    void visit(AST::Structure&) override;
+    void visit(AST::Variable&) override;
+    void visit(AST::Function&) override;
+    void visit(AST::Statement&) override;
+    void visit(AST::AssignmentStatement&) override;
+    void visit(AST::ReturnStatement&) override;
+    void visit(AST::Expression&) override;
+    void visit(AST::FieldAccessExpression&) override;
+    void visit(AST::IndexAccessExpression&) override;
+    void visit(AST::BinaryExpression&) override;
+    void visit(AST::IdentifierExpression&) override;
+    void visit(AST::TypeName&) override;
+
+private:
+    Type* infer(AST::Expression&);
+    Type* variable();
+    Type* resolve(AST::TypeName&);
+    void inferred(Type*);
+    void unify(Type*, Type*);
+
+    void substitute(Type*, SubstituionCallback&&);
+
+    AST::ShaderModule& m_shaderModule;
+    Type* m_inferredType { nullptr };
+    UnificationContext* m_unificationContext { nullptr };
+
+    // FIXME: remove this once we start allocating types
+    Type m_type;
+};
+
+class UnificationContext {
+public:
+    UnificationContext(TypeChecker* typeChecker)
+        : m_setForScope(typeChecker->m_unificationContext, this)
+    {
+    }
+
+    ~UnificationContext()
+    {
+        solveConstraints();
+        runSubstitutions();
+    }
+
+    void unify(Type* lhs, Type* rhs)
+    {
+        m_constraints.append({ lhs, rhs });
+    }
+
+    void substitute(Type* type, SubstituionCallback&& callback)
+    {
+        m_substitutions.append({ type, WTFMove(callback) });
+    }
+
+private:
+    void solveConstraints()
+    {
+        // FIXME: implement this
+    }
+
+    void runSubstitutions()
+    {
+        // FIXME: implement this
+    }
+
+    SetForScope<UnificationContext*> m_setForScope;
+    Vector<std::pair<Type*, Type*>> m_constraints;
+    Vector<std::pair<Type*, SubstituionCallback>> m_substitutions;
+};
+
+void TypeChecker::check()
+{
+    for (auto& structure : m_shaderModule.structures())
+        visit(structure);
+
+    for (auto& variable : m_shaderModule.variables())
+        visit(variable);
+
+    for (auto& function : m_shaderModule.functions())
+        visit(function);
+
+    for (auto& function : m_shaderModule.functions())
+        visit(function.body());
+}
+
+// Declarations
+void TypeChecker::visit(AST::Structure& structure)
+{
+    UnificationContext declarationContext(this);
+
+    // FIXME: allocate and build struct type from struct members
+    Type* structType = nullptr;
+    ContextProvider::introduceVariable(structure.name(), structType);
```

Comment:
Is this going to work in the presence of order-independent top-level declarations?

Reply:
Hum, we might have to do 2 passes (one to introduce the empty struct type and another to fill it)